proxy/stats_handler.go

```diff
@@ -0,0 +1,215 @@
+		return nil, nil, false
+	}
+
+	if !token.Valid {
+		return nil, nil, false
+	}
+
+	var hostKeys []string
+	var urls []string
+
+	if multiHost {
+		projectsOrServices, err := getProjectOrService(token)
+		if err != nil {
+			log.WithFields(log.Fields{"error": err}).Error("Error parsing token")
+			return nil, nil, false
+		}
+		for _, projectOrService := range projectsOrServices {
+			data := projectOrService
+			innerTokenString, ok := data["token"]
+			if !ok {
+				log.Error("Empty set of hosts or containers in project/service")
+				return nil, nil, false
+			}
+			innerJwtToken, err := parseRequestToken(innerTokenString, h.parsedPublicKey)
+			if err != nil {
+				log.WithFields(log.Fields{"error": err}).Error("Error parsing token")
+				return nil, nil, false
+			}
+			hostUuid, found := h.extractHostUuid(innerJwtToken)
+			if !found {
+				return nil, nil, false
+			}
+			hostKeys = append(hostKeys, hostUuid)
+			urlString, ok := data["url"]
+			if !ok {
+				log.WithFields(log.Fields{"error": err}).Error("Error parsing token")
+				return nil, nil, false
+			}
+			urlString = urlString + "?token=" + innerTokenString
+			urls = append(urls, urlString)
+		}
+	} else {
+		hostUuid, found := h.extractHostUuid(token)
+		if !found {
+			log.Error("could not find host uuid")
+			return nil, nil, false
+		}
+		hostKeys = append(hostKeys, hostUuid)
+		urls = append(urls, req.URL.String())
+	}
+	return hostKeys, urls, true
+}
+
+func getProjectOrService(token *jwt.Token) ([]map[string]string, error) {
+	data, ok := token.Claims["project"]
+	if !ok {
+		data, ok = token.Claims["service"]
+	}
+	if ok {
+		if interfaceList, isList := data.([]interface{}); isList {
+			projectList := []map[string]string{}
+			for _, inter := range interfaceList {
+				projectInterfaceMap, ok := inter.(map[string]interface{})
+				if ok {
+					projectMap := map[string]string{}
+					for key, value := range projectInterfaceMap {
+						valueString, ok := value.(string)
+						if ok {
+							projectMap[key] = valueString
+						} else {
+							return nil, fmt.Errorf("invalid project/service input data type")
+						}
+					}
+					projectList = append(projectList, projectMap)
+				} else {
+					return nil, fmt.Errorf("invalid project/service input data type")
+				}
+			}
+			return projectList, nil
+		}
+		return nil, fmt.Errorf("invalid project/service input data type")
+	}
+	return nil, fmt.Errorf("empty token")
+}
+
+func (h *StatsHandler) extractHostUuid(token *jwt.Token) (string, bool) {
+	hostUuid, found := token.Claims["hostUuid"]
+	if !found {
+		log.WithFields(log.Fields{"hostUuid": hostUuid}).Infof("HostUuid not found in token.")
+		return "", false
+	}
+	hostKey, ok := hostUuid.(string)
+	if !ok || !h.backend.hasBackend(hostKey) {
+		log.WithFields(log.Fields{"hostUuid": hostUuid}).Infof("Invalid HostUuid.")
+		return "", false
+	}
+	return hostKey, true
+}
+
+func (h *StatsHandler) closeConnection(ws *websocket.Conn) {
```

Comment:
Can you reuse the closeConnection function from frontend_handler? Just make it unbound function.